### PR TITLE
feat: (odk) Visualization decoration

### PR DESCRIPTION
**Pull request by lordmallam**
Adds 2 decorators to schemas generated from xforms with `select`, `select1` and `odk:rank`

Decorators:
`@aether_extended_type`: (Applied to all fields) A string indicating the original xform [type](https://docs.opendatakit.org/form-question-types)
`@aether_lookup`: (Available for only select fields) A list of `{ 'label': 'Foo', 'value': 'foo' }` representing all available options of the field.

Limitations: This does not apply to select fields linked to datasources using `itemset`. Research on-going on how to effectively resolve datasource references.

### CODE REVIEW

**Comment by admbtlr** [comment]
According to the ticket, the fields should be called `@aether_lookup` and `@aether_extended_type`. Can you change them please?

File: aether-odk-module/aether/odk/api/xform_utils.py [hunk 1]
```diff
@@ -282,8 +298,12 @@ def parse_xform_to_avro_schema(xml_definition, default_version=DEFAULT_XFORM_VER
             'name': current_name,
             'namespace': namespace,
             'doc': current_doc,
+            '@aether_extended_type': current_type,
         }
 
+        if select_options:
+            current_field['@aether_lookup'] = select_options if len(select_options) <= 20 else []
```
Comment (obdulia-losantos) [inline aether-odk-module/aether/odk/api/xform_utils.py]: `if` within `if`? why not the following?
```python
if select_options and len(select_options) <= 20:
    current_field['@aether_lookup'] = select_options
```

File: aether-odk-module/aether/odk/api/xform_utils.py [hunk 2]
```diff
@@ -271,6 +275,18 @@ def parse_xform_to_avro_schema(xml_definition, default_version=DEFAULT_XFORM_VER
         current_type = definition.get('type')
         current_name = xpath.split('/')[-1]
         current_doc = definition.get('label')
+        select_options = None
```
Comment (obdulia-losantos) [inline aether-odk-module/aether/odk/api/xform_utils.py]: why don't you include all of this inside the `__get_xform_instance_skeleton` method and return another property like `choices`?

File: aether-odk-module/aether/odk/api/xform_utils.py [hunk 3]
```diff
@@ -610,6 +617,9 @@ def __get_xform_instance_skeleton(xml_definition):
 
         - `label`, the linked label of the field, in case of multilanguage takes
           the translation for the default one.
+
+        - `choices`, a list of possible options if the field is of type select, select1, odk:rank
+          as { lable: 'Foo', value: 'foo'}
```
Comment (obdulia-losantos) [inline aether-odk-module/aether/odk/api/xform_utils.py]: typo?

File: aether-odk-module/aether/odk/api/xform_utils.py [hunk 4]
```diff
@@ -636,6 +646,20 @@ def __get_xform_instance_skeleton(xml_definition):
...
+                    select_options = list(select_node.get('item', []))
+                    # limitation: skips selects linked to a datasource with 'itemset'
+                    # todo: extend visualization decoration to itemsets
+                    for option in select_options:
+                        if 'label' in option and isinstance(option['label'], dict) \
+                                and '@ref' in option['label'] and option['label']['@ref'].startswith('jr:itext'):
+                            ref_id = option['label']['@ref'].split("'")[1]
+                            translated_label = itexts[ref_id]
+                            option['label'] = translated_label
+                    if select_options and len(select_options) <= 20:
```
Comment (obdulia-losantos) [inline aether-odk-module/aether/odk/api/xform_utils.py]: Could you change this magic number `20` with a constant like `SELECT_CHOICES_CUTOFF` of whatever you like most?

File: aether-odk-module/aether/odk/api/xform_utils.py [hunk 5]
```diff
@@ -636,6 +646,20 @@ def __get_xform_instance_skeleton(xml_definition):
                 xpath = bind_entry.get('@nodeset')
                 schema[xpath]['type'] = bind_entry.get('@type')
                 schema[xpath]['required'] = bind_entry.get('@required') == 'true()'
+                select_options = None
```
Comment (obdulia-losantos) [inline aether-odk-module/aether/odk/api/xform_utils.py]: You don't need this line.

File: aether-odk-module/aether/odk/api/tests/test_xform_utils.py [hunk 6]
```diff
@@ -25,6 +25,7 @@
     __get_xform_instance as get_instance,
     __get_xform_itexts as get_texts,
     __get_xform_label as get_label,
+    # __get_xform_select_options as get_select_options,
```
Comment (obdulia-losantos) [inline aether-odk-module/aether/odk/api/tests/test_xform_utils.py]: _cleaning the house_

File: aether-odk-module/aether/odk/api/xform_utils.py [hunk 7]
```diff
@@ -636,6 +646,20 @@ def __get_xform_instance_skeleton(xml_definition):
...
+                if schema[xpath]['type'] in SELECT_TAGS:
+                    select_node = list(__find_by_key_value(xform_dict, '@ref', xpath))[0]
+                    select_options = list(select_node.get('item', []))
+                    # limitation: skips selects linked to a datasource with 'itemset'
+                    # todo: extend visualization decoration to itemsets
+                    for option in select_options:
+                        if 'label' in option and isinstance(option['label'], dict) \
+                                and '@ref' in option['label'] and option['label']['@ref'].startswith('jr:itext'):
```
Comment (obdulia-losantos) [inline aether-odk-module/aether/odk/api/xform_utils.py]: simple is better

```python
if (
    isinstance(option.get('label'), dict) and
    option['label'].get('@ref', '').startswith('jr:itext')
):
```

**Review by obdulia-losantos — CHANGES_REQUESTED** [review]
Looks great, just a couple of small comments

**Review by obdulia-losantos — APPROVED** [review]
:clapper:

**Pull request merged**